git fetcher: fix detection of dirty git workspaces

mschwaig opened the pull request:
I used the steps outlined in #4140 to verify the issue was still there (using `nix flake metadata` which replaced `nix flake info`) with my local version of Nix and to verify that my first commit here fixes it.

The above did not fix the actual issue I was having similar to #5302, but in GitLab CI. There I think the problem is the code relying on the `.git/refs/heads` directory. I fixed this as well, but the error handling for `git rev-parse`is still incomplete. That's why this PR is a draft.

mschwaig commented:
I found a good source to consider for this PR:
https://github.com/git/git/blob/ad1260b6c994f7c0f9c259bd39f39979f7f4ecc2/git-sh-setup.sh

It has a `require_clean_work_tree` function that is meant as a building-block for automated scripts.

regnat reviewed (commented):
> but the error handling for git rev-parseis still incomplete

Is it in less complete than what was already there? It seems to me that this PR is handling the “not a git repository” case, which is already as much as what the other code was doing.

Do you mind also adding a test (well two) for that?

mschwaig commented:
> Is it in less complete than what was already there? It seems to me that this PR is handling the “not a git repository” case, which is already as much as what the other code was doing.

I did not have a good feeling about just throwing the error there, so I added a specific error message (daab386e9575705988dedbfe1548586da49ce7ae). The idea is that if you point nix at a broken repo it should tell you to fix it.

Next I will look into the commands and arguments called in that script I posted above, since I only found that after I had implemented my fixes and it seems like a good official source for how to do this right.

> Do you mind also adding a test (well two) for that?

I don't mind adding tests but I have to look into how to do that here and how to best test these changes. If you have suggestions for specific test cases in mind feel free to post them.

mschwaig commented:
I came to the conclusion that using `git diff` works better for our use case than using multiple lower-level commands like the previous version of my fix or [git-sh-setup.sh](https://github.com/git/git/blob/ad1260b6c994f7c0f9c259bd39f39979f7f4ecc2/git-sh-setup.sh) do. I also added tests for the two specific fixed issues.

regnat commented on src/libfetchers/git.cc:
```diff
@@ -223,22 +223,32 @@ struct GitInputScheme : InputScheme
+            } catch (ExecError & e) {
+                // git exits with status 128 here if it does not detect a repository.
+                if (!WIFEXITED(e.status) || WEXITSTATUS(e.status) != 128) {
+                    throw Error("Git tree '%s' is broken.\n'git rev-parse --verify HEAD' failed with exit code %d.", actualUrl, WEXITSTATUS(e.status));
```
Maybe use the `addTrace` mechanism so that the original exception is shown, just with more infos

```suggestion
                    e.addTrace({}, hintfmt("while trying to get the HEAD of the git tree '%s'", actualUrl);
                    throw;
```

regnat reviewed (changes requested):
Thanks, looks great now!

Just a small suggestion, but :+1: overall

mschwaig commented:
There were still a few issues with this PR before bc66c8316648a30b1e08e93d01a09f52333adcf4 with that git `rev-parse command`.

* It was writing some extra output when it failed.
* It was written to proceed when it does not detect a repo at all (which would still eventually lead to an error, but later).
* It also proceeded on repos without any commits, because not detecting a repo and detecting a repo without commits share the same status code. This was not done intentionally by me but seems to be expected behavior as the flake tests use lots of repos without commits.

Those cases are now distinguished by the error message `git rev-parse` returns. I added tests for a few of them. Doing this by status code is sadly not possible when repos without commits are allowed, at least not without adding more git commands.

I am not fully convinced that allowing repos without commits is a good feature. It looks like flakes kind of rely on this by `nix` moving `flake.nix` and `flake.lock` from the untracked state to the unstaged state, so that they are picked up by `git ls-files` and therefore part of a dirty tree right away without the user ever committing them even when there is no commit in the repo at all. I'm assuming this is done for convenience. I find it a bit magical. I'm not sure how much of this is intentionally part of the tests.

```
mschwaig@mutalisk ~/t/testAnotherFlakeThing6> git init
Initialized empty Git repository in /home/mschwaig/tmp/testAnotherFlakeThing6/.git/
mschwaig@mutalisk ~/t/testAnotherFlakeThing6 (main)> nix flake init
mschwaig@mutalisk ~/t/testAnotherFlakeThing6 (main)> git status
On branch main

No commits yet

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	new file:   flake.nix

no changes added to commit (use "git add" and/or "git commit -a")
mschwaig@mutalisk ~/t/testAnotherFlakeThing6 (main)> git ls-files
flake.nix
```

johnrichardrinehart commented:
Forgive my interjection: I'm affected by this PR via #5302. I see no problem with allowing fresh repos (no commits) to use flakes. In fact, I think that's a very valid use case for development purposes prior to an initial commit. The alternative is to `commit && echo "fix some stuff" && commit --amend` until everything is good - that UX is worse than the magical one (I agree it's kind of magical but it seems like it's origin is [here](https://github.com/NixOS/nix/blob/fcddaa4b9b22a910a24a513bc05579a12eb1db41/src/nix/flake.cc#L725-L729)).

Generally, though, in the case where someone is both:
1) starting a new repo
2) developing a flake

I think it's totally reasonable to expect that `nix` would need that `flake.nix` in the working tree and that, if it's not present, it will add it itself.

regnat commented on src/libfetchers/git.cc:
```diff
@@ -225,17 +225,25 @@ struct GitInputScheme : InputScheme
+
+            if (errorMessage.find("fatal: not a git repository") != std::string::npos) {
+                throw Error("'%s' is not a git repository.", actualUrl);
+            } else if (errorMessage.find("fatal: Needed a single revision") != std::string::npos) {
```
That’s not gonna work if the locale isn’t english because the error message will be translated:

```console
$ git init
$ nix flake init                                                 
$ LANG=C nix run github:mschwaig/nix/fix-git-workspace-dirty-detection -- build # Works with a C (or en) locale
warning: Git tree '/tmp/tmp.hjV72WecGo' is dirty
warning: creating lock file '/tmp/tmp.hjV72WecGo/flake.lock'
warning: Git tree '/tmp/tmp.hjV72WecGo' is dirty
$ LANG=fr_FR nix run github:mschwaig/nix/fix-git-workspace-dirty-detection -- build # But fails with another
error: Getting the HEAD of the git tree '/tmp/tmp.hjV72WecGo' failed with exit code 128:
       fatal: Une seule rision attendue
       
(use '--show-trace' to show detailed location information)
```

Maybe you could run another command here (`git rev-parse --all` ? `git rev-parse --git-dir` ? might be a better one) to distinguish between the no-repo and repo-but-no-commit cases. Since it’s an edge-case the cost of an extra command shouldn’t be an issue

mschwaig replied:
I have implemented a fix for this now in 0ef6df466f46a32f8c8677009303aea7d98ea2ee.

I have opted to explicitly setting the locale via an environment variable and still parsing the error messages, instead of running another `git` command, because I do not only want to distinguish the  no-repo and repo-but-no-commit cases, but also the case where `git` fails in some other unexpected error, which  should lead to fetch failing. I think looking at the error messages works better for that.

regnat reviewed (commented):
Thanks for the update, looking pretty great now.

Unfortunately the detection of a directory with no commit is still broken for non-english locales :/

(And to answer your question: I think it’s a useful thing to have it. I for one frequently create a git repository just to make it easy to filter out which files I want to have included in the flake by adding them or not to the index. Which is very handy and doesn’t require to create any commit)

mschwaig commented:
I have fixed and added a tests for that locale related failure, but I have also converted this to a draft for now.

My reasons are:
* The tests are not properly working yet (they don't fail if I undo 0ef6df466f46a32f8c8677009303aea7d98ea2ee - not sure yet what's wrong).
* I had to add a dependency to `glibcLocales` just for those tests which is a bit ugly.
* I originally wanted to reproduce the locale issue with `builtins.fetchGit`, but was unable to, so I'm using `nix falke show` now and put them with the flake tests.

Generally I'm not sure if it is worthwhile to add those tests.

mschwaig commented:
Since I did not get the tests for that locale-related failure to work I have removed them from this PR.

You can still find them at https://github.com/mschwaig/nix/tree/git-workspace-dirty-detection-locale-tests

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/flakes-embedding-git-commit-into-the-binary/17784/5

mschwaig commented:
I just rebased this onto current master, because I needed a more recent nix with those fixes.

mschwaig commented:
@thufschmitt can you or someone else take another look at this?

The `Changes requested` state here is outdated and it need another look, but I don't really know how to get out of that state.

mschwaig commented:
> Forgive my interjection: I'm affected by this PR via #5302. I see no problem with allowing fresh repos (no commits) to use flakes. In fact, I think that's a very valid use case for development purposes prior to an initial commit. The alternative is to `commit && echo "fix some stuff" && commit --amend` until everything is good - that UX is worse than the magical one (I agree it's kind of magical but it seems like its origin is [here](https://github.com/NixOS/nix/blob/fcddaa4b9b22a910a24a513bc05579a12eb1db41/src/nix/flake.cc#L725-L729)).
> 
> Generally, though, in the case where someone is both:
> 
>     1. starting a new repo
> 
>     2. developing a flake
> 
> 
> I think it's totally reasonable to expect that `nix` would need that `flake.nix` in the working tree and that, if it's not present, it will add it itself.
> 
> My apologies if I'm misunderstanding something and/or if my input isn't wanted, here.

@johnrichardrinehart I agree. I came to the same conclusion, so this PR does not change that behaviour. Thanks for the input.

thufschmitt reviewed (commented):
Thanks for the reminder @mschwaig .

This is all a bit dirty − esp the reliance of the error message from Git − but I guess we don’t have much choice (well, except maybe switching to libgit2, but that would be a much wider scope thing to do).

So :+1: from me (provided the CI agrees). Just leaving it open a couple extra days in case @edolstra wants to have another look
